README.txt
Introduction
============

collective.portlet.feedmixer is a portlet for `Plone 3`_ which can show one
or multiple content feeds.

Feedmixer uses feedparser_ to process feeds which allows it to support RSS
0.90, Netscape RSS 0.91, Userland RSS 0.91, RSS 0.92, RSS 0.93, RSS 0.94,
RSS 1.0, RSS 2.0, Atom 0.3, Atom 1.0, and CDF feeds

.. _Plone 3: http://plone.org/
.. _feedparser: http://www.feedparser.org/


Installing
==========

This package requires Plone 3.0 or later and feedparser.

Installing without buildout
---------------------------

First install feedparser. You can install this in either the system
python path or in the lib/python directory of your Zope instance. If you
have setuptools installed you can do this using easy_install::

  easy_install FeedParser

If you do not have setuptools you can install it manually using the setup.py
script in the feedparser source. If you want to install feedparser inside
your Zope instance instead of system wide you can its ''--prefix='' option
to install in the ''lib/python'' directory of your Zope instance.

Next you need to install this package. This can also be done by installing
it in either your system path packages or in the lib/python directory of
your Zope instance. As with feedparser you can do this using either
easy_install or via the setup.py script.

After installing the package it needs to be registered in your Zope
instance.  This can be done by putting a
collective.portlet.feedmixer-configure.zcml file in the etc/pakage-includes
directory with this content::

  <include package="collective.portlet.feedmixer" />

or, alternatively, you can add that line to the configure.zcml in a
package or Product that is already registered.

Installing with buildout
------------------------

If you are using `buildout`_ to manage your instance installing
collective.portlet.feedmixer is even simpler. You can install it by adding
it to the eggs line for your instance::

  [instance]
  eggs = collective.portlet.feedmixer
  zcml = collective.portlet.feedmixer

The last line tells buildout to generate a zcml snippet that tells Zope
to configure collective.portlet.feedmixer.

If another package depends on the feedmixer egg or includes its zcml
directly you do not need to specify anything in the buildout configuration:
buildout will detect this automatically.

After updating the configuration you need to run the ''bin/buildout'', which
will take care of updating your system.

.. _buildout: http://pypi.python.org/pypi/zc.buildout


Using the portlet
=================

In order to use the feedmixer portlet you will first need to install the
feedmixer product in your site. This is done through the standard ''Add-on
products'' control panel in the Plone site setup. After this has been done
a new ''Feed Mixer' portlet type will be available.

Feedmixer portlets have several configurable options:

Portlet Title
  This will be used as the title of the portlet.

Maximum time to cache feed data
  This determins how long feeds can be cached before feedmixer will attempt
  to refresh the feed data. In order to reduce the load on other servers
  it is advisable to set this to a reasonably large value.

Number of items to display
  How many items to show in the portlet.

URL(s) for all feeds
  A list of URLs for all feeds. All standard RSS and Atom feeds are
  supported as well as CDF feeds.


Performance
===========

The portlet tries to be very friendly and will only reload a feed if required.
It does this in several different ways.

All feeds are stored in a shared cache as defined in `plone.memoize`_. This
defaults to being a Zope RAMCache but you can choose different caches by
registering a different ''ICacheChooser'' utility. The cache key used
is ''collective.portlet.feedmixer.FeedCache''. For example for deployments with
multiple Zope instances you can use a memcached_ based cache which will
allow you to share downloaded feeds amongst all Zope instances.

If a feed entry has expired in the cache it needs to be reloaded. However
the feedmixer portlet will try to use both `ETAGs` and HTTP Last-Modified
headers to check if the feed has been updated since the last time it was
downloaded. If it has not been updated the timeout for the cached feed will
be updated and continue to be used.

.. _plone.memoize: http://pypi.python.org/pypi/plone.memoize
.. _ETAGs: http://en.wikipedia.org/wiki/HTTP_ETag
.. _memcached: http://www.danga.com/memcached/


Copyright and credits
=====================

collective.portlet.feedmixer is copyright 2007 by `Jarn`_ (formerly known
as Plone Solutions).

It was written by Wichert Akkerman.

.. _Jarn: http://www.jarn.com/




HISTORY.txt
Changelog
=========

Release 1.0
-----------

* Initial release.



collective/portlet/feedmixer/tests/__init__.py
import os

home = __path__[0]

FEED_ONE = os.path.join(home, "feed1.xml")



collective/portlet/feedmixer/tests/test_feed.py
import unittest
from zope.interface.verify import verifyObject
from collective.portlet.feedmixer.interfaces import IFeedMixer
from collective.portlet.feedmixer.portlet import Assignment

class ConstructionTests(unittest.TestCase):
    def testInterface(self):
        feed=Assignment()
        verifyObject(IFeedMixer, feed)

    def testParameterOrder(self):
        feed=Assignment("title", "feeds", "items_shown", "cache_timeout")
        self.assertEqual(feed.title, "title")
        self.assertEqual(feed.feeds, "feeds")
        self.assertEqual(feed.items_shown, "items_shown")
        self.assertEqual(feed.cache_timeout, "cache_timeout")

    def testNamerParameters(self):
        feed=Assignment(title="title", feeds="feeds",
                items_shown="items_shown", cache_timeout="cache_timeout")
        self.assertEqual(feed.title, "title")
        self.assertEqual(feed.feeds, "feeds")
        self.assertEqual(feed.items_shown, "items_shown")
        self.assertEqual(feed.cache_timeout, "cache_timeout")

def test_suite():
    suite=unittest.TestSuite()
    suite.addTest(unittest.makeSuite(ConstructionTests))
    return suite


collective/portlet/feedmixer/portlet.py
import itertools
import time
import feedparser

from zope.interface import implements
from zope.component import getMultiAdapter
from zope.component import getUtility

from plone.app.portlets.portlets import base
from plone.memoize.instance import memoize
from plone.memoize.interfaces import ICacheChooser

from zope.formlib import form
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

from collective.portlet.feedmixer.interfaces import IFeedMixer



class Assignment(base.Assignment):
    """Portlet assignment.
    
    This is what is actually managed through the portlets UI and associated
    with columns.
    """
    implements(IFeedMixer)

    title = u"Feed Viewer"
    feeds = u""
    items_shown = 5
    cache_timeout = 900

    def __init__(self, title=title, feeds=feeds, items_shown=items_shown,
            cache_timeout=cache_timeout):
        self.title=title
        self.feeds=feeds
        self.items_shown=items_shown
        self.cache_timeout=cache_timeout
        
    @property        
    def feed_urls(self):
        return (url.strip() for url in self.feeds.split())
        

    def Title(self):
        """Returns the title. The function is used by Plone to render <title> correctly."""
        return self.title
        

    def cleanFeed(self, feed):
        """Sanitize the feed.

        This function makes sure all feed and entry data we depend on us
        present and in proper form.
        """
        for entry in feed.entries:
            entry["feed"]=feed.feed
            if not "published_parsed" in entry:
                entry["published_parsed"]=entry["updated_parsed"]
                entry["published"]=entry["updated"]



    def getFeed(self, url):
        """Fetch a feed.

        This may return a cached result if the cache entry is considered to
        be fresh. Returned feeds have been cleaned using the cleanFeed method.
        """
        now=time.time()

        chooser=getUtility(ICacheChooser)
        cache=chooser("collective.portlet.feedmixer.FeedCache")

        cached_data=cache.get(url, None)
        if cached_data is not None:
            (timestamp, feed)=cached_data
            if now-timestamp<self.cache_timeout:
                return feed

            newfeed=feedparser.parse(url,
                    etag=getattr(feed, "etag", None),
                    modified=getattr(feed, "modified", None))
            if newfeed.status==304:
                self.cleanFeed(feed)
                cache[url]=(now+self.cache_timeout, feed)
                return feed

        feed=feedparser.parse(url)
        self.cleanFeed(feed)
        cache[url]=(now+self.cache_timeout, feed)

        return feed


    def mergeEntriesFromFeeds(self, feeds):
        if not feeds:
            return []
        if len(feeds)==1:
            return feeds[0].entries

        entries=list(itertools.chain(*(feed.entries for feed in feeds)))
        entries.sort(key=lambda x: x["published_parsed"], reverse=True)

        return entries

    
    @memoize
    def entries(self):
        feeds=[self.getFeed(url) for url in self.data.feed_urls]
        feeds=[feed for feed in feeds if feed is not None]
        entries=self.mergeEntriesFromFeeds(feeds)
        return entries


class Renderer(base.Renderer):
    """Portlet renderer.
    """
    render = ViewPageTemplateFile("portlet.pt")

    @property
    def available(self):
        return bool(self.data.entries())

    @property
    def title(self):
        return self.data.title

    @property
    def entries(self):
        return self.data.entries()[:self.data.items_shown]

    @property
    def more_url(self):
        state=getMultiAdapter((self.context, self.request), name="plone_context_state")
        folder=state.folder()
        return "%s/++contextportlets++%s/%s/full_feed" % \
                (folder.absolute_url(),
                 self.manager.__name__,
                 self.data.__name__)



class AddForm(base.AddForm):
    """Portlet add form.
    """
    form_fields = form.Fields(IFeedMixer)

    def create(self, data):
        return Assignment(**data)


class EditForm(base.EditForm):
    """Portlet edit form.
    """
    form_fields = form.Fields(IFeedMixer)


collective/portlet/feedmixer/feedmixer.py
import itertools
import time
import feedparser

from zope.interface import implements
from zope.component import getMultiAdapter
from zope.component import getUtility

from plone.app.portlets.portlets import base
from plone.memoize.instance import memoize
from plone.memoize.interfaces import ICacheChooser

from zope.formlib import form
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

from collective.portlet.feedmixer.interfaces import IFeedMixer



class Assignment(base.Assignment):
    """Portlet assignment.
    
    This is what is actually managed through the portlets UI and associated
    with columns.
    """
    implements(IFeedMixer)

    title = u"Feed Viewer"
    feeds = u""
    items_shown = 5
    cache_timeout = 900

    def __init__(self, title=title, feeds=feeds, items_shown=items_shown,
            cache_timeout=cache_timeout):
        self.title=title
        self.feeds=feeds
        self.items_shown=items_shown
        self.cache_timeout=cache_timeout
        
    @property        
    def feed_urls(self):
        return (url.strip() for url in self.feeds.split())
        

    def cleanFeed(self, feed):
        """Sanitize the feed.

        This function makes sure all feed and entry data we depend on us
        present and in proper form.
        """
        for entry in feed.entries:
            if not "published_parsed" in entry:
                entry["published_parsed"]=entry["updated_parsed"]
                entry["published"]=entry["updated"]



    def getFeed(self, url):
        """Fetch a feed.

        This may return a cached result if the cache entry is considered to
        be fresh. Returned feeds have been cleaned using the cleanFeed method.
        """
        now=time.time()

        chooser=getUtility(ICacheChooser)
        cache=chooser("collective.portlet.feedmixer.FeedCache")

        cached_data=cache.get(url, None)
        if cached_data is not None:
            (timestamp, feed)=cached_data
            if now-timestamp<self.cache_timeout:
                return feed

            newfeed=feedparser.parse(url,
                    etag=getattr(feed, "etag", None),
                    modified=getattr(feed, "modified", None))
            if newfeed.status==304:
                self.cleanFeed(feed)
                cache[url]=(now+self.cache_timeout, feed)
                return feed

        feed=feedparser.parse(url)
        self.cleanFeed(feed)
        cache[url]=(now+self.cache_timeout, feed)

        return feed


    def mergeEntriesFromFeeds(self, feeds):
        if not feeds:
            return []
        if len(feeds)==1:
            return feeds[0].entries

        entries=list(itertools.chain(*(feed.entries for feed in feeds)))
        entries.sort(key=lambda x: x["published_parsed"], reverse=True)

        return entries

    
    @memoize
    def entries(self):
        feeds=[self.getFeed(url) for url in self.data.feed_urls]
        feeds=[feed for feed in feeds if feed is not None]
        entries=self.mergeEntriesFromFeeds(feeds)
        return entries


class Renderer(base.Renderer):
    """Portlet renderer.
    """
    render = ViewPageTemplateFile("feedmixer.pt")

    @property
    def available(self):
        return bool(self.data.entries)

    @property
    def title(self):
        return self.data.title

    @property
    def entries(self):
        return self.data.entries[:self.data.items_shown]

    @property
    def more_url(self):
        state=getMultiAdapter((self.context, self.request), name="plone_context_state")
        folder=state.folder()
        return "%s/++contextportlets++%s/%s/full_feed" % \
                (folder.absolute_url(),
                 self.manager.__name__,
                 self.data.__name__)



class AddForm(base.AddForm):
    """Portlet add form.
    """
    form_fields = form.Fields(IFeedMixer)

    def create(self, data):
        return Assignment(**data)


class EditForm(base.EditForm):
    """Portlet edit form.
    """
    form_fields = form.Fields(IFeedMixer)


collective/portlet/feedmixer/interfaces.py
from plone.portlets.interfaces import IPortletDataProvider
from collective.portlet.feedmixer import FeedMixerMessageFactory as _
from zope import schema
from Products.validation import validation

def is_url_list(data):
    verify=validation.validatorFor("isURL")
    for url in (x.strip() for x in data.split()):
        if verify(url)!=True:
            return False
    return True




class IFeedMixer(IPortletDataProvider):
    """A portlet which aggregates multiple feeds.
    """
    title = schema.TextLine(
            title=_(u"heading_title",
                default=u"Portlet Title"),
            description=_(u"description_title",
                default=u""),
            default=u"",
            required=True)

    cache_timeout = schema.Choice(
            title=_(u"heading_cache_timeout",
                default=u"Maximum time to cache feed data"),
            description=_(u"description_cache_timeout",
                default=u""),
            default=900,
            required=True,
            vocabulary="collective.portlet.feedmixer.timeouts")


    items_shown = schema.Int(
            title=_(u"heading_items_shown",
                default=u"Number of items to display"),
            description=_(u"description_items_shown",
                default=u""),
            default=5,
            required=True)

    feeds = schema.ASCII(
            title=_(u"heading_feeds",
                default=u"URL(s) for all feeds"),
            description=_(u"description_feeds",
                default=u"Enter the URLs for all feeds here, one URL per "
                        u"line. RSS 0.9x, RSS 1.0, RSS 2.0, CDF, Atom 0.3 "
                        u"and ATOM 1.0 feeds are supported."),
            required=True,
            constraint=is_url_list)

    def entries():
        """Return feed entries for all feeds.

        The entries from all feeds will be combined in a single listed and
        sorted by publication date.
        """

